### PR TITLE
VideoPress Block: edit video settings while uploading

### PR DESCRIPTION
**Pull request by pgk**
Allows users to edit some video settings while the video is uploading. Users can set the poster image and change the title.

#### Changes proposed in this Pull Request:

* Adds a new screen on video upload for editing video settings.
* Upload Poster tab
* Select poster from frame tab.
* Edit title text field.

#### Jetpack product discussion

pxWta-1pr-p2

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* D81302-code to be applied too if it is not already merged. In that case, public api needs to be sandboxed.
* Create a new post and upload videos. You see an expandable element for editing video settings.
Screenshot 2022-05-31 at 5 25 31 PM

* Expand, and try different combinations of  changing the title, uploading a poster, or selecting a frame as a poster on each video.
* Pressing "Done" should bring up the video embed. Verify that the settings you updated took effect.
* Uploading a video without expanding the editor should work the same way it did.

Screenshot 2022-05-31 at 5 22 14 PM
Screenshot 2022-05-31 at 5 21 56 PM
Screenshot 2022-05-31 at 5 20 37 PM
Screenshot 2022-05-31 at 5 20 19 PM
Screenshot 2022-05-31 at 5 20 08 PM

### CODE REVIEW

**Comment by jeffgolenski** [comment]
@pgk Hey panos! Thanks for taking this one! This is a cool feature. 

I think for accessibility reasons, we should have some clear headlines (even though you give explanation in the description). Could we add "Edit your title" & "Select a poster frame" as small headlines above each option? Also, these are optional, right? Perhaps we should add "(optional)" in the description as well. 

From a visual standpoint, I wonder if people will miss the chevron to expand. Would it be worth expanding by default to prompt more usage?

**Comment by pgk** [comment]
> @pgk Hey panos! Thanks for taking this one! This is a cool feature.
> 
> I think for accessibility reasons, we should have some clear headlines (even though you give explanation in the description). Could we add "Edit your title" & "Select a poster frame" as small headlines above each option? Also, these are optional, right? Perhaps we should add "(optional)" in the description as well.
> 
> From a visual standpoint, I wonder if people will miss the chevron to expand. Would it be worth expanding by default to prompt more usage?

Cool ideas! cc: @evilluendas who is going to be creating the designs for these

File: projects/plugins/jetpack/modules/videopress/js/gutenberg-video-upload.js [hunk 1]
```diff
@@ -1,9 +1,47 @@
...
+		} );
+	} );
+};
+
 window.videoPressUploadTrack = function ( guid, kind, srcLang, label, vttFile ) {
 	// eslint-disable-next-line no-undef
 	return new Promise( function ( resolve, reject ) {
 		wp.media
-			.ajax( 'videopress-get-upload-token', { async: true, data: { filename: vttFile.name } } )
+			.ajax( 'videopress-get-upload-token', { async: true, data: { filename: vttFile.name } } ) // todo: maybe remove filename from here (not needed)
```
Comment (thedebian) [inline projects/plugins/jetpack/modules/videopress/js/gutenberg-video-upload.js:44]: Should it be removed ? (or should the comment be removed ? 😄 )

Comment (jgcaruso) [reply projects/plugins/jetpack/modules/videopress/js/gutenberg-video-upload.js:44]: I remember writing this `todo:` when mobbing. I think it should be removed but it should also be tested first before committing to make sure its not actually needed.

**Comment by jgcaruso** [comment]
The title doesn't appear to be saved even though the request succeeds. Looks like its probably a race condition with the transcoder. Similar problem seems to be happening with the frame selection... my selected frame doesn't get used.

Not entirely sure about how to solve this, but wanted to document it somewhere.

**Comment by roundhill** [comment]
I have this PR updated to the new design:

Screen Shot 2022-06-30 at 7 09 36 PM

But stuff is still broken 😉 If someone wants to pick this up while I'm AFK, here's what I've logged as remaining tasks:

- [x] The actual setting of the poster image doesn't work, even with D81302-code applied to my sandbox (could be a CORS error?)
- [x] You can't completely empty out the `Title` field, if you get to 0 characters it reverts back to the original state of the text.
- [x] I haven't tested on touch / mobile yet.

**Comment by thedebian** [comment]
>     * [x]  The actual setting of the poster image doesn't work, even with D81302-code applied to my sandbox (could be a CORS error?)
There's a mix between a CORS problem, and another problem in wpcom where we try to respond with the poster while its generation is delayed.
I asked for help for the CORS problem, as it's related to a recent change in a header Jetpack is sending, and I'll update D81302-code for the wpcom problem.

**Comment by roundhill** [comment]
- [ ] Per slack convo, let's have the block always do the `Done` button behavior so the UX isn't confusing.

**Comment by roundhill** [comment]
Selecting a frame is working fine for me now, but not selecting a custom poster. Title is also still not saving... @thedebian were you still working on that? Maybe we should also mob on this again to finish it up 😅

**Comment by thedebian** [comment]
> Selecting a frame is working fine for me now, but not selecting a custom poster. Title is also still not saving... @thedebian were you still working on that? Maybe we should also mob on this again to finish it up 😅

Both selecting a frame and a custom poster should" work" (if both wpcom PRs are applied), but the race condition is still there;
Title is using a /v2 endpoint which only seems to set the video meta, but not update the attachment properly, that's what I was trying to figure out, but wasn't able to make it work before the weekend (trying to replace the /wpcom/v2 call with a public-api call (the same used on Calypso, I'm mainly missing the site id and probably couple of other details).
I'll get back to it on Monday, but mobbing that could be nice too ! :)

**Comment by thedebian** [comment]
Title update should be fixed now !

**Review by roundhill — COMMENTED** [review]
I noticed something while testing and pushed a fix, the play button wasn't centered after selecting a custom poster image. 

In general I approve and thing this is ready to go but someone else should review my latest changes :D

**Review by thedebian — DISMISSED** [review]
Looks good after peer review

**Comment by samiff** [comment]
This is a nice upgrade! Here are some things I noticed while testing:

1. When I upload a video, and before it has finished uploading completely (showing the Done button), if I try selecting a custom poster image using the slider, then the upload seems to stay stuck at 100% and I never get the Done button showing.
2. It wasn't immediately clear to me that I have to click the "Done" button after selecting a custom poster image using the slider. If I just save the post, then the video block will still be in the post but not with the poster image I selected.
3. In my opinion, the slider bar seems a bit "hidden" and it's small to interact with. Looks like in earlier screenshots there was a small circle for dragging? I don't see that in the current iteration.
4. Would it be possible to update the block settings in the sidebar when setting a custom poster image from the block?

Markup on 2022-08-03 at 11:52:22

**Comment by thedebian** [comment]
Hi @samiff !

>     1. When I upload a video, and before it has finished uploading completely (showing the Done button), if I try selecting a custom poster image using the slider, then the upload seems to stay stuck at 100% and I never get the Done button showing.
I think that's a general Tus quirk, I'll spend a bit of time trying to look into it, but it's hard to reproduce it consistently.

>     2. It wasn't immediately clear to me that I have to click the "Done" button after selecting a custom poster image using the slider. If I just save the post, then the video block will still be in the post but not with the poster image I selected.
Awesome suggestion, I just updated the PR to add this 👍 

>     3. In my opinion, the slider bar seems a bit "hidden" and it's small to interact with. Looks like in earlier screenshots there was a small circle for dragging? I don't see that in the current iteration.
I think that's a design iteration that has been done, maybe @evilluendas can answer that (and provide us a nice solution that we could implement, or no solution if it's fine).

>     4. Would it be possible to update the block settings in the sidebar when setting a custom poster image from the block?
I'll try to update the PR with this change too 👍

**Review by samiff — DISMISSED** [review]
@thedebian Those two new changes work great for me 👍 

> I think that's a general Tus quirk, I'll spend a bit of time trying to look into it, but it's hard to reproduce it consistently.

I didn't notice it today while I was testing, so I'm wondering if it was more to do with uploading the same video file more than once and the resumable portion just getting confused.

**Comment by bindlegirl** [comment]
@pgk This PR has some conflicts that need resolving and is ready to get merged.

**Review by samiff — APPROVED** [review]

**Pull request merged**